include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,414 @@
+/* Copyright 2013-2018 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
+ *                     Klaus Steiniger, Felix Schmitt, Benjamin Worpitz,
+ *                     Juncheng E
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
+#include "picongpu/plugins/ISimulationPlugin.hpp"
+#include "picongpu/plugins/common/stringHelpers.hpp"
+#include "picongpu/plugins/saxs/Saxs.kernel"
+
+#include <pmacc/dataManagement/DataConnector.hpp>
+#include <pmacc/dimensions/DataSpaceOperations.hpp>
+#include <pmacc/mappings/kernel/AreaMapping.hpp>
+#include <pmacc/mpi/MPIReduce.hpp>
+#include <pmacc/mpi/reduceMethods/Reduce.hpp>
+#include <pmacc/nvidia/functors/Add.hpp>
+#include <pmacc/traits/GetNumWorkers.hpp>
+#include <pmacc/traits/HasIdentifier.hpp>
+
+#include <boost/filesystem.hpp>
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+namespace picongpu
+{
+using namespace pmacc;
+
+namespace po = boost::program_options;
+
+/** SAXS plugin
+ * This SAXS plugin simulates the SAXS scattering pattern on-the-fly
+ * from the particle positions obtained from PIConGPU.
+ **/
+
+template <class ParticlesType> class Saxs : public ISimulationPlugin
+{
+  private:
+    using SuperCellSize = MappingDesc::SuperCellSize;
+
+    //! The real part of structure factor
+    GridBuffer<float1_64, DIM1> *sumfcoskr;
+    //! The imaginary part of structure factor
+    GridBuffer<float1_64, DIM1> *sumfsinkr;
+    //! Number of real particles
+    GridBuffer<float1_64, DIM1> *np;
+    //! Number of macro particles
+    GridBuffer<int64_t, DIM1> *nmp;
+
+    MappingDesc *cellDescription;
+    std::string notifyPeriod;
+    std::string speciesName;
+    std::string pluginName;
+    std::string pluginPrefix;
+    std::string filename_prefix;
+
+    /** Range of scattering vector
+     * The scattering vecotor here is defined as
+     * 4*pi*sin(theta)/lambda, where 2theta is the angle
+     * between scattered and incident beam
+     **/
+    float3_64 q_min, q_max, q_step;
+    //! Number of scattering vectors
+    unsigned int n_qx, n_qy, n_qz, n_q;
+
+    float1_64 *sumfcoskr_master;
+    float1_64 *sumfsinkr_master;
+    float1_64 *intensity_master;
```

Comment:
I think arrays `sumfcoskr_master`, `sumfsinkr_master` and `intensity_master` are a little better as `std::vector`s. Then you can get a plain pointer, e.g. for reduction, with calling `.data()` method.
Also, not sure why these are `float1_64` and not just `float_64`; but not against it if there are reasons.